src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewFactory.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Tagging;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Editor.Tagging;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Responsible for wiring up all the different components of the document outline feature.
+    /// </summary>
+    internal static class DocumentOutlineViewFactory
+    {
+        public static DocumentOutlineView CreateView(
+            ILanguageServiceBroker2 languageServiceBroker,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            threadingContext.ThrowIfNotOnUIThread();
+            var (textViewEventSource, textBuffer) = CreateEventSource(asyncListener, editorAdaptersFactoryService, codeWindow);
+            var viewModel = new DocumentOutlineViewModel(languageServiceBroker, asyncListener, textViewEventSource, textBuffer, threadingContext);
+            return new DocumentOutlineView(viewModel, editorAdaptersFactoryService, codeWindow);
+        }
+
+        private static (CompilationAvailableTaggerEventSource, ITextBuffer) CreateEventSource(
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
```

Comment:
can you add threading asserts *everything*.

Note: this is not me being paranoid.  We literally just had an issue with the editor where they called into us on the wrong thread, and it was a threading assert that caught it.  Even they are not immune to mistakes here, and the asserts really help us ensure everyone is behaving exactly as we expect.